File: solutions/lua/perfect-numbers/1/perfect-numbers.lua
-- factors compute all factors of n
-- a factor of n is any number greater than 0 and lower than n which is a perfectdivisor of n
local function factors(n)

  if n == 1 then return {} end

  local fs = { 1 }
  for f = 2, n - 1 do
    if n % f == 0 then
      table.insert(fs, f)
    end
  end

  return fs

end

local function aliquot_sum(n)

  local res = 0
  for _, v in ipairs(factors(n)) do
    res = res + v
  end

  return res

end

local function classify(n)

  local sum = aliquot_sum(n)
  if sum == n then
    return 'perfect'
  elseif sum > n then
    return 'abundant'
  else
    return 'deficient'
  end
end

return {
  aliquot_sum = aliquot_sum,
  classify = classify
}

File: solutions/javascript/perfect-numbers/1/perfect-numbers.js

// Computes all the factors of n.
// A factor of n is any number lower than n which is a perfect divisor.
function factors(n) {
  
    if (n ==1) return [];
  
  let factors = [1];
    for (let f = 2; f < n; f++) {
        if (n % f == 0) factors.push(f)
    }
    return factors;
}

// Returns the sum of the elements of an array.
const sum = (array) => array.reduce((acc,val) => acc+val, 0); 

// Classifies n as perfact, abundant or deficient.
export function classify(n) {
  
  if (n <= 0 || Math.trunc(n) != n) {
    throw new Error('Classification is only possible for natural numbers.');
  }

  let aliquotSum = sum(factors(n));

  if (aliquotSum == n) return 'perfect';
  if (aliquotSum > n) return 'abundant';
  return 'deficient';
};
